File: nodejs/src/library/protocol.ts
import { Buffer } from "node:buffer";
import {
  LIVE_STATUS_PROPERTIES,
  YEELIGHT_METHODS,
  type LiveStatusProperty,
  type YeelightAdjustAction,
  type YeelightAdjustProperty,
  type YeelightCommandParamsMap,
  type YeelightDiscoveredDevice,
  type YeelightEffect,
  type YeelightFlowAction,
  type YeelightFlowTuple,
  type YeelightMethod,
  type YeelightNotification,
  type YeelightProgressFlowOptions,
  type YeelightPowerMode,
  type YeelightPropertyValues,
  type YeelightScene,
  type YeelightSceneParams
} from "./types";

export const DISCOVERY_ADDRESS = "239.255.255.250";
export const DISCOVERY_PORT = 1982;
export const DEFAULT_DISCOVERY_TIMEOUT_MS = 3000;
export const DEFAULT_COMMAND_TIMEOUT_MS = 3000;
export const DEFAULT_EFFECT: YeelightEffect = "sudden";
export const DEFAULT_DURATION_MS = 30;
export const DEFAULT_CONTROL_PORT = 55443;
export const DEFAULT_CACHE_TTL_MS = 10 * 60 * 1000;

const METHOD_SET = new Set<string>(YEELIGHT_METHODS);

export function isYeelightMethod(value: string): value is YeelightMethod {
  return METHOD_SET.has(value);
}

export function buildDiscoveryRequest(): Buffer {
  return Buffer.from(
    [
      "M-SEARCH * HTTP/1.1",
      `HOST: ${DISCOVERY_ADDRESS}:${DISCOVERY_PORT}`,
      'MAN: "ssdp:discover"',
      "MX: 2",
      "ST: wifi_bulb",
      "",
      ""
    ].join("\r\n"),
    "utf8"
  );
}

export function parseDiscoveryHeaders(headers: Record<string, string>): YeelightDiscoveredDevice | null {
  const normalizedHeaders = Object.entries(headers || {}).reduce<Record<string, string>>(
    (accumulator, [key, value]) => {
      accumulator[String(key).toLowerCase()] = String(value).trim();
      return accumulator;
    },
    {}
  );

  if (!normalizedHeaders.location) {
    return null;
  }

  const locationMatch = /^yeelight:\/\/([^:]+):(\d+)$/i.exec(normalizedHeaders.location);
  if (!locationMatch) {
    return null;
  }

  const host = locationMatch[1];
  const portText = locationMatch[2];
  if (!host || !portText) {
    return null;
  }

  const port = Number.parseInt(portText, 10);
  const support = (normalizedHeaders.support || "")
    .split(/\s+/)
    .map((item) => item.trim())
    .filter(Boolean);

  return {
    id: normalizedHeaders.id || `${host}:${port}`,
    host,
    port,
    model: normalizedHeaders.model || "",
    firmwareVersion: normalizedHeaders.fw_ver || "",
    name: decodeDeviceName(normalizedHeaders.name || ""),
    power: normalizedHeaders.power || "",
    brightness: normalizedHeaders.bright || "",
    colorMode: normalizedHeaders.color_mode || "",
    colorTemperature: normalizedHeaders.ct || "",
    rgb: normalizedHeaders.rgb || "",
    hue: normalizedHeaders.hue || "",
    saturation: normalizedHeaders.sat || "",
    support,
    location: normalizedHeaders.location
  };
}

export function parseDiscoveryResponse(message: Buffer | string): YeelightDiscoveredDevice | null {
  const lines = String(message)
    .split(/\r?\n/)
    .filter(Boolean);
  if (!lines.length) {
    return null;
  }

  const headers: Record<string, string> = {};
  let lastKey: string | null = null;

  for (const line of lines) {
    const separatorIndex = line.indexOf(":");
    if (separatorIndex === -1) {
      if (lastKey && line.trim()) {
        headers[lastKey] = `${headers[lastKey]} ${line.trim()}`.trim();
      }
      continue;
    }

    const key = line.slice(0, separatorIndex).trim().toLowerCase();
    const value = line.slice(separatorIndex + 1).trim();
    headers[key] = value;
    lastKey = key;
  }

  return parseDiscoveryHeaders(headers);
}

export function decodeDeviceName(value: string): string {
  const trimmed = value.trim();
  if (!trimmed) {
    return "";
  }

  if (!/^[A-Za-z0-9+/=]+$/.test(trimmed) || trimmed.length % 4 !== 0) {
    return trimmed;
  }

  const decoded = Buffer.from(trimmed, "base64").toString("utf8");
  if (!decoded || /[\u0000-\u0008\u000E-\u001F]/.test(decoded)) {
    return trimmed;
  }

  const normalizedInput = trimmed.replace(/=+$/, "");
  const normalizedOutput = Buffer.from(decoded, "utf8").toString("base64").replace(/=+$/, "");
  return normalizedInput === normalizedOutput ? decoded : trimmed;
}

export function buildCommand<TMethod extends YeelightMethod>(
  method: TMethod,
  params: YeelightCommandParamsMap[TMethod],
  id: number
): string {
  return `${JSON.stringify({ id, method, params })}\r\n`;
}

export function splitSocketBuffer(pending: string, chunk: string): { lines: string[]; pending: string } {
  const combined = `${pending}${chunk}`;
  const parts = combined.split("\r\n");
  const nextPending = parts.pop() || "";
  const lines = parts.map((line) => line.trim()).filter(Boolean);
  return { lines, pending: nextPending };
}

export function normalizeStatus<TProperty extends string>(
  properties: readonly TProperty[],
  result: readonly string[]
): YeelightPropertyValues<TProperty> {
  return properties.reduce<YeelightPropertyValues<TProperty>>((accumulator, propertyName, index) => {
    accumulator[propertyName] = result[index] ?? null;
    return accumulator;
  }, {} as YeelightPropertyValues<TProperty>);
}

export function parseNotificationMessage(message: string): YeelightNotification | null {
  const parsed = JSON.parse(message) as Partial<YeelightNotification>;
  if (parsed.method !== "props" || typeof parsed.params !== "object" || !parsed.params) {
    return null;
  }

  const normalizedParams = Object.entries(parsed.params).reduce<YeelightNotification["params"]>(
    (accumulator, [key, value]) => {
      accumulator[key as keyof YeelightNotification["params"]] = String(value);
      return accumulator;
    },
    {}
  );

  return {
    method: "props",
    params: normalizedParams
  };
}

export function buildFlowExpression(tuples: readonly YeelightFlowTuple[]): string {
  if (tuples.length === 0) {
    throw new Error("At least one flow tuple is required.");
  }

  return tuples
    .map((tuple) => {
      validateFlowTuple(tuple);
      return [tuple.duration, tuple.mode, tuple.value, tuple.brightness].join(",");
    })
    .join(",");
}

export function buildProgressFlowTuples(
  rgbValue: number,
  options: YeelightProgressFlowOptions = {}
): YeelightFlowTuple[] {
  validateRgbValue(rgbValue);

  const stepDuration = options.stepDuration ?? 180;
  const pauseDuration = options.pauseDuration ?? 260;
  const minBrightness = options.minBrightness ?? 12;
  const maxBrightness = options.maxBrightness ?? 100;
  const steps = options.steps ?? 5;

  validateDuration(stepDuration, 50);
  validateDuration(pauseDuration, 50);
  validateBrightness(minBrightness);
  validateBrightness(maxBrightness);

  if (!Number.isInteger(steps) || steps < 2) {
    throw new Error("Progress flow steps must be an integer greater than or equal to 2.");
  }

  if (minBrightness >= maxBrightness) {
    throw new Error("Progress flow minBrightness must be lower than maxBrightness.");
  }

  const tuples: YeelightFlowTuple[] = [];
  const brightnessRange = maxBrightness - minBrightness;

  for (let index = 0; index < steps; index += 1) {
    const brightness =
      index === steps - 1
        ? maxBrightness
        : Math.round(minBrightness + (brightnessRange * index) / (steps - 1));

    tuples.push({
      duration: stepDuration,
      mode: 1,
      value: rgbValue,
      brightness
    });
  }

  tuples.push({
    duration: pauseDuration,
    mode: 7,
    value: 0,
    brightness: 0
  });

  return tuples;
}

export function buildProgressFlowExpression(
  rgbValue: number,
  options: YeelightProgressFlowOptions = {}
): string {
  return buildFlowExpression(buildProgressFlowTuples(rgbValue, options));
}

export function serializeScene(scene: YeelightScene): YeelightSceneParams {
  switch (scene.class) {
    case "color":
      validateRgbValue(scene.rgbValue);
      validateBrightness(scene.brightness);
      return ["color", scene.rgbValue, scene.brightness];
    case "hsv":
      validateHue(scene.hue);
      validateSaturation(scene.saturation);
      validateBrightness(scene.brightness);
      return ["hsv", scene.hue, scene.saturation, scene.brightness];
    case "ct":
      validateColorTemperature(scene.ctValue);
      validateBrightness(scene.brightness);
      return ["ct", scene.ctValue, scene.brightness];
    case "cf":
      validateFlowCount(scene.count);
      validateFlowAction(scene.action);
      return [
        "cf",
        scene.count,
        scene.action,
        typeof scene.flowExpression === "string"
          ? scene.flowExpression
          : buildFlowExpression(scene.flowExpression)
      ];
    case "auto_delay_off":
      validateBrightness(scene.brightness);
      validatePositiveInteger(scene.minutes, "Auto-delay minutes");
      return ["auto_delay_off", scene.brightness, scene.minutes];
  }
}

export function validateEffect(effect: string): asserts effect is YeelightEffect {
  if (effect !== "sudden" && effect !== "smooth") {
    throw new Error('Effect must be "sudden" or "smooth".');
  }
}

export function validateDuration(duration: number, minimum = 30): void {
  if (!Number.isInteger(duration) || duration < minimum) {
    throw new Error(`Duration must be an integer greater than or equal to ${minimum} milliseconds.`);
  }
}

export function validateBrightness(brightness: number): void {
  if (!Number.isInteger(brightness) || brightness < 1 || brightness > 100) {
    throw new Error("Brightness must be an integer between 1 and 100.");
  }
}

export function validateColorTemperature(colorTemperature: number): void {
  if (!Number.isInteger(colorTemperature) || colorTemperature < 1700 || colorTemperature > 6500) {
    throw new Error("Color temperature must be an integer between 1700 and 6500 Kelvin.");
  }
}

export function validateRgbValue(rgbValue: number): void {
  if (!Number.isInteger(rgbValue) || rgbValue < 0 || rgbValue > 0xffffff) {
    throw new Error("RGB color must be an integer between 0 and 16777215.");
  }
}

export function validateRgbComponent(component: number, label: string): void {
  if (!Number.isInteger(component) || component < 0 || component > 255) {
    throw new Error(`${label} must be an integer between 0 and 255.`);
  }
}

export function validateHue(hue: number): void {
  if (!Number.isInteger(hue) || hue < 0 || hue > 359) {
    throw new Error("Hue must be an integer between 0 and 359.");
  }
}

export function validateSaturation(saturation: number): void {
  if (!Number.isInteger(saturation) || saturation < 0 || saturation > 100) {
    throw new Error("Saturation must be an integer between 0 and 100.");
  }
}

export function validatePowerMode(mode: number): asserts mode is YeelightPowerMode {
  if (!Number.isInteger(mode) || mode < 0 || mode > 5) {
    throw new Error("Power mode must be an integer between 0 and 5.");
  }
}

export function validatePercentage(percentage: number, label = "Percentage"): void {
  if (!Number.isInteger(percentage) || percentage < -100 || percentage > 100) {
    throw new Error(`${label} must be an integer between -100 and 100.`);
  }
}

export function validateName(name: string): void {
  if (!name.trim()) {
    throw new Error("Device name must not be empty.");
  }
  if (Buffer.byteLength(name, "utf8") > 64) {
    throw new Error("Device name must be 64 bytes or fewer.");
  }
}

export function validateCronType(type: number): void {
  if (type !== 0) {
    throw new Error("Cron type must be 0 according to the Yeelight spec.");
  }
}

export function validateAdjustAction(action: string): asserts action is YeelightAdjustAction {
  if (action !== "increase" && action !== "decrease" && action !== "circle") {
    throw new Error('Adjust action must be "increase", "decrease", or "circle".');
  }
}

export function validateAdjustProperty(property: string): asserts property is YeelightAdjustProperty {
  if (property !== "bright" && property !== "ct" && property !== "color") {
    throw new Error('Adjust property must be "bright", "ct", or "color".');
  }
}

export function validateFlowAction(action: number): asserts action is YeelightFlowAction {
  if (!Number.isInteger(action) || action < 0 || action > 2) {
    throw new Error("Flow action must be 0, 1, or 2.");
  }
}

export function validateFlowCount(count: number): void {
  if (!Number.isInteger(count) || count < 0) {
    throw new Error("Flow count must be an integer greater than or equal to 0.");
  }
}

export function validateFlowTuple(tuple: YeelightFlowTuple): void {
  validateDuration(tuple.duration, 50);
  if (tuple.mode !== 1 && tuple.mode !== 2 && tuple.mode !== 7) {
    throw new Error("Flow tuple mode must be 1 (rgb), 2 (ct), or 7 (sleep).");
  }

  if (tuple.mode === 1) {
    validateRgbValue(tuple.value);
  } else if (tuple.mode === 2) {
    validateColorTemperature(tuple.value);
  }

  if (tuple.mode !== 7 && tuple.brightness !== -1) {
    validateBrightness(tuple.brightness);
  }
}

export function validateMusicStart(host: string, port: number): void {
  if (!host.trim()) {
    throw new Error("Music mode host must not be empty.");
  }
  validatePositiveInteger(port, "Music mode port");
}

export function validatePositiveInteger(value: number, label: string): void {
  if (!Number.isInteger(value) || value <= 0) {
    throw new Error(`${label} must be a positive integer.`);
  }
}

export function formatDeviceLabel(device: Pick<YeelightDiscoveredDevice, "name" | "id" | "host" | "port">): string {
  const title = device.name || device.id || `${device.host}:${device.port}`;
  return `${title} (${device.host}:${device.port})`;
}

export function formatDeviceSummary(device: YeelightDiscoveredDevice): string {
  const support = device.support.length ? device.support.join(", ") : "unknown";
  const name = device.name || "(unnamed)";
  return [
    `${name} - ${device.id}`,
    `  host: ${device.host}:${device.port}`,
    `  model: ${device.model || "unknown"}`,
    `  power: ${device.power || "unknown"}`,
    `  brightness: ${device.brightness || "unknown"}`,
    `  support: ${support}`
  ].join("\n");
}

export function toRgbValue(red: number, green: number, blue: number): number {
  validateRgbComponent(red, "Red");
  validateRgbComponent(green, "Green");
  validateRgbComponent(blue, "Blue");
  return (red << 16) | (green << 8) | blue;
}

export function formatRgbHex(rgbValue: number): string {
  validateRgbValue(rgbValue);
  return `#${rgbValue.toString(16).padStart(6, "0").toUpperCase()}`;
}

export { LIVE_STATUS_PROPERTIES };
export type { LiveStatusProperty };

File: rust/src/protocol.rs
use std::collections::BTreeMap;

use base64::Engine;
use serde_json::{Value, json};

use crate::error::{Result, YeelightError};
use crate::types::{
    FlowExpression, LIVE_STATUS_PROPERTIES, YEELIGHT_METHODS, YeelightDiscoveredDevice,
    YeelightFlowTuple, YeelightNotification, YeelightPropertyValues, YeelightScene,
};

pub const DISCOVERY_ADDRESS: &str = "239.255.255.250";
pub const DISCOVERY_PORT: u16 = 1982;
pub const DEFAULT_DISCOVERY_TIMEOUT_MS: u64 = 3000;
pub const DEFAULT_COMMAND_TIMEOUT_MS: u64 = 3000;
pub const DEFAULT_EFFECT: &str = "sudden";
pub const DEFAULT_DURATION_MS: u64 = 30;
pub const DEFAULT_CONTROL_PORT: u16 = 55443;
pub const DEFAULT_CACHE_TTL_MS: u64 = 10 * 60 * 1000;

pub fn is_yeelight_method(value: &str) -> bool {
    YEELIGHT_METHODS.contains(&value)
}

pub fn build_discovery_request() -> Vec<u8> {
    [
        "M-SEARCH * HTTP/1.1",
        &format!("HOST: {DISCOVERY_ADDRESS}:{DISCOVERY_PORT}"),
        "MAN: \"ssdp:discover\"",
        "MX: 2",
        "ST: wifi_bulb",
        "",
        "",
    ]
    .join("\r\n")
    .into_bytes()
}

pub fn parse_discovery_headers(headers: &BTreeMap<String, String>) -> Option<YeelightDiscoveredDevice> {
    let normalized: BTreeMap<String, String> = headers
        .iter()
        .map(|(key, value)| (key.to_ascii_lowercase(), value.trim().to_owned()))
        .collect();

    let location = normalized.get("location")?;
    let remainder = location.strip_prefix("yeelight://")?;
    let (host, port_text) = remainder.rsplit_once(':')?;
    if host.is_empty() {
        return None;
    }

    let port = port_text.parse::<u16>().ok()?;
    let support = normalized
        .get("support")
        .map(|items| {
            items
                .split_whitespace()
                .map(ToOwned::to_owned)
                .collect::<Vec<_>>()
        })
        .unwrap_or_default();

    Some(YeelightDiscoveredDevice {
        id: normalized
            .get("id")
            .cloned()
            .unwrap_or_else(|| format!("{host}:{port}")),
        host: host.to_owned(),
        port,
        model: normalized.get("model").cloned().unwrap_or_default(),
        firmware_version: normalized.get("fw_ver").cloned().unwrap_or_default(),
        name: decode_device_name(normalized.get("name").map(String::as_str).unwrap_or("")),
        power: normalized.get("power").cloned().unwrap_or_default(),
        brightness: normalized.get("bright").cloned().unwrap_or_default(),
        color_mode: normalized.get("color_mode").cloned().unwrap_or_default(),
        color_temperature: normalized.get("ct").cloned().unwrap_or_default(),
        rgb: normalized.get("rgb").cloned().unwrap_or_default(),
        hue: normalized.get("hue").cloned().unwrap_or_default(),
        saturation: normalized.get("sat").cloned().unwrap_or_default(),
        support,
        location: location.clone(),
    })
}

pub fn parse_discovery_response(message: impl AsRef<[u8]>) -> Option<YeelightDiscoveredDevice> {
    let raw = String::from_utf8_lossy(message.as_ref());
    let mut headers = BTreeMap::new();
    let mut last_key: Option<String> = None;

    for line in raw
        .split('\n')
        .map(|line| line.trim_end_matches('\r'))
        .filter(|line| !line.trim().is_empty())
    {
        if let Some(separator_index) = line.find(':') {
            let key = line[..separator_index].trim().to_ascii_lowercase();
            let value = line[separator_index + 1..].trim().to_owned();
            headers.insert(key.clone(), value);
            last_key = Some(key);
            continue;
        }

        if let Some(key) = &last_key {
            if let Some(existing) = headers.get_mut(key) {
                *existing = format!("{} {}", existing, line.trim()).trim().to_owned();
            }
        }
    }

    parse_discovery_headers(&headers)
}

pub fn decode_device_name(value: &str) -> String {
    let trimmed = value.trim();
    if trimmed.is_empty() {
        return String::new();
    }

    if !trimmed
        .bytes()
        .all(|byte| matches!(byte, b'A'..=b'Z' | b'a'..=b'z' | b'0'..=b'9' | b'+' | b'/' | b'='))
        || trimmed.len() % 4 != 0
    {
        return trimmed.to_owned();
    }

    let decoded = match base64::engine::general_purpose::STANDARD.decode(trimmed) {
        Ok(bytes) => bytes,
        Err(_) => return trimmed.to_owned(),
    };

    let decoded = match String::from_utf8(decoded) {
        Ok(value) => value,
        Err(_) => return trimmed.to_owned(),
    };

    if decoded.chars().any(|character| matches!(character as u32, 0x00..=0x08 | 0x0E..=0x1F)) {
        return trimmed.to_owned();
    }

    let normalized_input = trimmed.trim_end_matches('=');
    let normalized_output = base64::engine::general_purpose::STANDARD
        .encode(decoded.as_bytes())
        .trim_end_matches('=')
        .to_owned();
    if normalized_input == normalized_output {
        decoded
    } else {
        trimmed.to_owned()
    }
}

pub fn build_command(method: &str, params: &[Value], id: u64) -> String {
    format!("{}\r\n", json!({ "id": id, "method": method, "params": params }))
}

pub fn split_socket_buffer(pending: &str, chunk: &str) -> (Vec<String>, String) {
    let combined = format!("{pending}{chunk}");
    let mut parts: Vec<&str> = combined.split("\r\n").collect();
    let next_pending = parts.pop().unwrap_or_default().to_owned();
    let lines = parts
        .into_iter()
        .map(str::trim)
        .filter(|line| !line.is_empty())
        .map(ToOwned::to_owned)
        .collect();
    (lines, next_pending)
}

pub fn normalize_status(properties: &[&str], result: &[String]) -> YeelightPropertyValues {
    properties
        .iter()
        .enumerate()
        .map(|(index, property)| ((*property).to_owned(), result.get(index).cloned()))
        .collect()
}

pub fn parse_notification_message(message: &str) -> Result<Option<YeelightNotification>> {
    let parsed: Value = serde_json::from_str(message)?;
    let Value::Object(root) = parsed else {
        return Ok(None);
    };

    let Some(method) = root.get("method").and_then(Value::as_str) else {
        return Ok(None);
    };
    if method != "props" {
        return Ok(None);
    }

    let Some(Value::Object(params)) = root.get("params") else {
        return Ok(None);
    };

    let params = params
        .iter()
        .map(|(key, value)| (key.clone(), value_as_string(value)))
        .collect();

    Ok(Some(YeelightNotification {
        method: "props".to_owned(),
        params,
    }))
}

pub fn build_flow_expression(tuples: &[YeelightFlowTuple]) -> Result<String> {
    if tuples.is_empty() {
        return Err(YeelightError::from("At least one flow tuple is required."));
    }

    let mut parts = Vec::with_capacity(tuples.len());
    for tuple in tuples {
        validate_flow_tuple(tuple)?;
        parts.push(format!(
            "{},{},{},{}",
            tuple.duration, tuple.mode, tuple.value, tuple.brightness
        ));
    }
    Ok(parts.join(","))
}

pub fn serialize_scene(scene: &YeelightScene) -> Result<Vec<Value>> {
    match scene {
        YeelightScene::Color {
            rgb_value,
            brightness,
        } => {
            validate_rgb_value(*rgb_value)?;
            validate_brightness(*brightness)?;
            Ok(vec![json!("color"), json!(rgb_value), json!(brightness)])
        }
        YeelightScene::Hsv {
            hue,
            saturation,
            brightness,
        } => {
            validate_hue(*hue)?;
            validate_saturation(*saturation)?;
            validate_brightness(*brightness)?;
            Ok(vec![json!("hsv"), json!(hue), json!(saturation), json!(brightness)])
        }
        YeelightScene::Ct {
            ct_value,
            brightness,
        } => {
            validate_color_temperature(*ct_value)?;
            validate_brightness(*brightness)?;
            Ok(vec![json!("ct"), json!(ct_value), json!(brightness)])
        }
        YeelightScene::Cf {
            count,
            action,
            flow_expression,
        } => {
            validate_flow_count(*count)?;
            validate_flow_action(*action)?;
            let flow = match flow_expression {
                FlowExpression::Serialized(value) => value.clone(),
                FlowExpression::Tuples(tuples) => build_flow_expression(tuples)?,
            };
            Ok(vec![json!("cf"), json!(count), json!(action), json!(flow)])
        }
        YeelightScene::AutoDelayOff {
            brightness,
            minutes,
        } => {
            validate_brightness(*brightness)?;
            validate_positive_integer(*minutes, "Auto-delay minutes")?;
            Ok(vec![json!("auto_delay_off"), json!(brightness), json!(minutes)])
        }
    }
}

pub fn validate_effect(effect: &str) -> Result<()> {
    match effect {
        "sudden" | "smooth" => Ok(()),
        _ => Err(YeelightError::from("Effect must be \"sudden\" or \"smooth\".")),
    }
}

pub fn validate_duration(duration: u64, minimum: u64) -> Result<()> {
    if duration < minimum {
        return Err(YeelightError::from(format!(
            "Duration must be an integer greater than or equal to {minimum} milliseconds."
        )));
    }
    Ok(())
}

pub fn validate_brightness(brightness: u8) -> Result<()> {
    if !(1..=100).contains(&brightness) {
        return Err(YeelightError::from("Brightness must be an integer between 1 and 100."));
    }
    Ok(())
}

pub fn validate_color_temperature(color_temperature: u16) -> Result<()> {
    if !(1700..=6500).contains(&color_temperature) {
        return Err(YeelightError::from(
            "Color temperature must be an integer between 1700 and 6500 Kelvin.",
        ));
    }
    Ok(())
}

pub fn validate_rgb_value(rgb_value: u32) -> Result<()> {
    if rgb_value > 0x00ff_ffff {
        return Err(YeelightError::from(
            "RGB color must be an integer between 0 and 16777215.",
        ));
    }
    Ok(())
}

pub fn validate_rgb_component(component: u16, label: &str) -> Result<()> {
    if component > 255 {
        return Err(YeelightError::from(format!(
            "{label} must be an integer between 0 and 255."
        )));
    }
    Ok(())
}

pub fn validate_hue(hue: u16) -> Result<()> {
    if hue > 359 {
        return Err(YeelightError::from("Hue must be an integer between 0 and 359."));
    }
    Ok(())
}

pub fn validate_saturation(saturation: u8) -> Result<()> {
    if saturation > 100 {
        return Err(YeelightError::from(
            "Saturation must be an integer between 0 and 100.",
        ));
    }
    Ok(())
}

pub fn validate_power_mode(mode: u8) -> Result<()> {
    if mode > 5 {
        return Err(YeelightError::from(
            "Power mode must be an integer between 0 and 5.",
        ));
    }
    Ok(())
}

pub fn validate_percentage(percentage: i16) -> Result<()> {
    if !(-100..=100).contains(&percentage) {
        return Err(YeelightError::from(
            "Percentage must be an integer between -100 and 100.",
        ));
    }
    Ok(())
}

pub fn validate_name(name: &str) -> Result<()> {
    if name.trim().is_empty() {
        return Err(YeelightError::from("Device name must not be empty."));
    }
    if name.len() > 64 {
        return Err(YeelightError::from("Device name must be 64 bytes or fewer."));
    }
    Ok(())
}

pub fn validate_cron_type(value: u8) -> Result<()> {
    if value != 0 {
        return Err(YeelightError::from(
            "Cron type must be 0 according to the Yeelight spec.",
        ));
    }
    Ok(())
}

pub fn validate_adjust_action(action: &str) -> Result<()> {
    match action {
        "increase" | "decrease" | "circle" => Ok(()),
        _ => Err(YeelightError::from(
            "Adjust action must be \"increase\", \"decrease\", or \"circle\".",
        )),
    }
}

pub fn validate_adjust_property(property: &str) -> Result<()> {
    match property {
        "bright" | "ct" | "color" => Ok(()),
        _ => Err(YeelightError::from(
            "Adjust property must be \"bright\", \"ct\", or \"color\".",
        )),
    }
}

pub fn validate_flow_action(action: u8) -> Result<()> {
    if action > 2 {
        return Err(YeelightError::from("Flow action must be 0, 1, or 2."));
    }
    Ok(())
}

pub fn validate_flow_count(_count: u64) -> Result<()> {
    Ok(())
}

pub fn validate_flow_tuple(tuple: &YeelightFlowTuple) -> Result<()> {
    validate_duration(tuple.duration, 50)?;
    match tuple.mode {
        1 => validate_rgb_value(tuple.value)?,
        2 => validate_color_temperature(tuple.value as u16)?,
        7 => {}
        _ => {
            return Err(YeelightError::from(
                "Flow tuple mode must be 1 (rgb), 2 (ct), or 7 (sleep).",
            ));
        }
    }

    if tuple.mode != 7 && tuple.brightness != -1 {
        if !(1..=100).contains(&tuple.brightness) {
            return Err(YeelightError::from("Brightness must be an integer between 1 and 100."));
        }
    }
    Ok(())
}

pub fn validate_music_start(host: &str, port: u16) -> Result<()> {
    if host.trim().is_empty() {
        return Err(YeelightError::from("Music mode host must not be empty."));
    }
    validate_positive_integer(u64::from(port), "Music mode port")
}

pub fn validate_positive_integer(value: u64, label: &str) -> Result<()> {
    if value == 0 {
        return Err(YeelightError::from(format!("{label} must be a positive integer.")));
    }
    Ok(())
}

pub fn format_device_label(device: &YeelightDiscoveredDevice) -> String {
    let title = if !device.name.is_empty() {
        device.name.clone()
    } else if !device.id.is_empty() {
        device.id.clone()
    } else {
        format!("{}:{}", device.host, device.port)
    };
    format!("{title} ({}:{})", device.host, device.port)
}

pub fn format_device_summary(device: &YeelightDiscoveredDevice) -> String {
    let support = if device.support.is_empty() {
        "unknown".to_owned()
    } else {
        device.support.join(", ")
    };
    let name = if device.name.is_empty() {
        "(unnamed)"
    } else {
        device.name.as_str()
    };

    [
        format!("{name} - {}", device.id),
        format!("  host: {}:{}", device.host, device.port),
        format!("  model: {}", if device.model.is_empty() { "unknown" } else { &device.model }),
        format!("  power: {}", if device.power.is_empty() { "unknown" } else { &device.power }),
        format!(
            "  brightness: {}",
            if device.brightness.is_empty() { "unknown" } else { &device.brightness }
        ),
        format!("  support: {support}"),
    ]
    .join("\n")
}

pub fn to_rgb_value(red: u16, green: u16, blue: u16) -> Result<u32> {
    validate_rgb_component(red, "Red")?;
    validate_rgb_component(green, "Green")?;
    validate_rgb_component(blue, "Blue")?;
    Ok(((red as u32) << 16) | ((green as u32) << 8) | blue as u32)
}

pub fn format_rgb_hex(rgb_value: u32) -> Result<String> {
    validate_rgb_value(rgb_value)?;
    Ok(format!("#{rgb_value:06X}"))
}

pub fn live_status_properties() -> &'static [&'static str] {
    LIVE_STATUS_PROPERTIES
}

fn value_as_string(value: &Value) -> String {
    match value {
        Value::String(inner) => inner.clone(),
        Value::Null => "null".to_owned(),
        Value::Bool(inner) => inner.to_string(),
        Value::Number(inner) => inner.to_string(),
        Value::Array(_) | Value::Object(_) => value.to_string(),
    }
}
